File: python/src/threads/permissions/shell.py
"""The shell parser behind bash rules.

A command is tokenized with POSIX quoting and split into simple commands on `;`, `&&`, `||`,
`|`, `&`, newlines and the edges of subshells and substitutions (`(`, `)`, backticks). Reserved
words (`if`, `then`, `do`, `!`, ...) that open a simple command are read past, so deny and ask
rules see every command at any depth. It only makes rules conservative: anything it can't read
is flagged unparseable, which can be denied but never allowed. The sandbox stays the security
boundary.
"""

import re
from dataclasses import dataclass

_SEPARATORS = ("&&", "||", ";", "|", "&", "\n")
# Leading assignments that change what the command runs: such a command never matches allow.
_DANGEROUS_ENV = frozenset({"PATH", "BASH_ENV", "ENV", "IFS", "PYTHONPATH", "NODE_OPTIONS", "PS4"})
_ASSIGNMENT = re.compile(r"([A-Za-z_][A-Za-z0-9_]*)=")
_UNPARSEABLE_WORDS = frozenset({"eval", "exec", "function"})
# Reserved words that can open a simple command without being it.
_KEYWORDS = frozenset(
    {
        "!",
        "{",
        "}",
        "if",
        "then",
        "elif",
        "else",
        "fi",
        "while",
        "until",
        "do",
        "done",
        "esac",
        "coproc",
    }
)
# Unquoted characters that expand, redirect, group or comment: a command holding one never
# matches allow.
_META = frozenset("$<>(){}`#")
# Unquoted characters that open or close a subshell or a substitution: they end a simple command.
_EDGES = frozenset("()`")
_RAW_WORD = re.compile(r"[^\s;&|()`$<>{}\"'\\]+")


@dataclass(frozen=True, slots=True)
class Simple:
    """One simple command after stripping leading assignments and wrappers."""

    words: tuple[str, ...]
    allowable: bool
    """False when a dangerous environment assignment was stripped from it, or when it opens with or
    follows a reserved word."""


@dataclass(frozen=True, slots=True)
class Shell:
    commands: tuple[Simple, ...]
    unparseable: bool
    plain: bool
    """No unquoted metacharacter: no escape, expansion, redirection, group or comment (separators
    don't count). Only a plain command can match an allow rule (it fails closed)."""
    raw_words: tuple[str, ...]
    """The raw text split on whitespace and shell punctuation, for denying input that isn't
    plain."""


def parse(text: str) -> Shell:
    lexer = _Lexer(text)
    tokens, unparseable, meta = lexer.run()
    firsts = [simple(t) for t in tokens]
    unparseable = unparseable or any(_unreadable(c.words) for c in firsts)
    commands = tuple(c for first in firsts for c in _unwrap_keywords(first))
    return Shell(commands, unparseable, not meta, tuple(_RAW_WORD.findall(text)))


def _unwrap_keywords(command: Simple) -> tuple[Simple, ...]:
    """A command that opens with a reserved word, and the command after the word, both as deny
    and ask rules read them. Neither is allowed."""
    if not command.words:
        return ()
    if command.words[0] not in _KEYWORDS:
        return (command,)
    inner = _unwrap_keywords(simple(command.words[1:]))
    return tuple(Simple(c.words, allowable=False) for c in (command, *inner))


def _unreadable(words: tuple[str, ...]) -> bool:
    # Brace groups, function definitions, eval and exec.
    head = words[0] if words else ""
    return head in _UNPARSEABLE_WORDS or "{" in words or "}" in words


def simple(tokens: tuple[str, ...]) -> Simple:
    words = list(tokens)
    allowable = True
    while words:
        assigned = _ASSIGNMENT.match(words[0])
        if assigned is not None:
            name = assigned.group(1)
            allowable = allowable and not _dangerous(name)
            del words[0]
            continue
        wrapped = _wrapper_length(words)
        if wrapped == 0:
            break
        del words[:wrapped]
    return Simple(tuple(words), allowable)


def _dangerous(name: str) -> bool:
    return name in _DANGEROUS_ENV or name.startswith(("LD_", "DYLD_"))


def _wrapper_length(words: list[str]) -> int:
    """How many leading words are a `timeout N`, `nice`, `nohup` or `time` wrapper."""
    head = words[0]
    if head in ("nohup", "time"):
        return 1
    if head not in ("timeout", "nice"):
        return 0
    n = 1
    while n < len(words) and words[n].startswith("-"):
        # -s SIGNAL, -k DURATION and nice's -n N take a separate argument.
        n += 2 if words[n] in ("-s", "-k", "-n") else 1
    return n + 1 if head == "timeout" else n


class _Lexer:
    def __init__(self, text: str) -> None:
        self._text = text
        self._i = 0
        self._word: list[str] = []
        self._has_word = False
        self._tokens: list[str] = []
        self._commands: list[tuple[str, ...]] = []
        self._unparseable = False
        self._meta = False

    def run(self) -> tuple[list[tuple[str, ...]], bool, bool]:
        while self._i < len(self._text):
            self._step(self._text[self._i])
        self._end_command()
        return self._commands, self._unparseable, self._meta

    def _step(self, c: str) -> None:
        if c in " \t":
            self._end_word()
            self._i += 1
        elif c == "'":
            self._single()
        elif c == '"':
            self._double()
        elif c == "\\":
            self._meta = True
            self._take(self._text[self._i + 1 : self._i + 2].replace("\n", ""))
            self._i += 2
        elif c in _EDGES:
            # Subshells, function definitions, command and process substitution: the command
            # inside is read on its own.
            self._unparseable = self._meta = True
            self._end_command()
            self._i += 1
        elif not self._separator():
            if self._text.startswith("<<", self._i):
                self._unparseable = True  # a here-doc
            self._meta = self._meta or c in _META
            self._take(c)
            self._i += 1

    def _separator(self) -> bool:
        for sep in _SEPARATORS:
            if self._text.startswith(sep, self._i):
                self._end_command()
                self._i += len(sep)
                return True
        return False

    def _single(self) -> None:
        end = self._text.find("'", self._i + 1)
        if end < 0:
            self._unparseable = True
            end = len(self._text)
        self._take(self._text[self._i + 1 : end])
        self._i = end + 1

    def _double(self) -> None:
        self._i += 1
        self._has_word = True
        while self._i < len(self._text) and self._text[self._i] != '"':
            c = self._text[self._i]
            if c == "`" or self._text.startswith("$(", self._i):
                self._unparseable = True
            self._meta = self._meta or c in "\\$`"
            if c == "\\" and self._text[self._i + 1 : self._i + 2] in ('"', "\\", "$", "`"):
                self._i += 1
                c = self._text[self._i]
            self._take(c)
            self._i += 1
        if self._i >= len(self._text):
            self._unparseable = True
        self._i += 1

    def _take(self, chars: str) -> None:
        self._word.append(chars)
        self._has_word = True

    def _end_word(self) -> None:
        if self._has_word:
            self._tokens.append("".join(self._word))
        self._word, self._has_word = [], False

    def _end_command(self) -> None:
        self._end_word()
        self._commands.append(tuple(self._tokens))
        self._tokens = []

File: typescript/packages/core/src/permissions/shell.ts
// shell commands tokenized for rule matching only. The sandbox stays the
// security boundary; this parser only has to make rules conservative.

/** One simple command, as rules see it. */
export type SimpleCommand = {
  readonly tokens: readonly string[];
  /**
   * False after a dangerous leading assignment, or when the command opens with or follows a
   * reserved word (`if`, `do`, `!`, ...): such a command never matches an allow rule.
   */
  readonly allowable: boolean;
};

/**
 * A command line split into simple commands. An unparseable line holds a construct the lexer
 * doesn't model: its commands are a best effort that deny and ask rules still read, its raw
 * `words` are read too, and no allow rule matches it.
 */
export type ParsedShell = {
  readonly commands: readonly SimpleCommand[];
  readonly unparseable: boolean;
  readonly words: readonly string[];
};

const DANGEROUS_ENV =
  /^(PATH|LD_\w*|DYLD_\w*|BASH_ENV|ENV|IFS|PYTHONPATH|NODE_OPTIONS|PS4)$/;
const ASSIGNMENT = /^([A-Za-z_]\w*)=/;
// Words of an unparseable command: what a deny rule may still find.
const RAW_WORD = /[^\s;&|()<>`$"'{}\\]+/g;
// Reserved words that can open a simple command without being it. The command after one is
// read too, so a deny rule sees `rm` in `if true; then rm -rf /; fi`.
const KEYWORDS = new Set([
  "!",
  "{",
  "}",
  "if",
  "then",
  "elif",
  "else",
  "fi",
  "while",
  "until",
  "do",
  "done",
  "esac",
  "coproc",
]);
const WRAPPERS = new Set(["timeout", "nice", "nohup", "time"]);
// Wrapper options whose value is the next word: timeout's signal and kill-after, nice's niceness.
const OPTIONS_WITH_VALUE = new Set([
  "-s",
  "-k",
  "-n",
  "--signal",
  "--kill-after",
  "--adjustment",
]);
// Commands whose words run as code the parser doesn't read.
const OPAQUE = new Set(["eval", "exec", "function"]);
// One lexer step: blanks, a separator, a modelled word piece (bare text, a single-quoted string,
// or a double-quoted one holding no escape or expansion), a group edge (a subshell's or a
// substitution's parenthesis, a backtick), or anything else (a double-quoted string with an
// escape or expansion, an escape, or one character such as `$`, `>`, `{` or `#`). A group edge
// ends the simple command, so the command inside is read on its own; a group edge or anything
// else makes the line unparseable.
const PIECE =
  /([ \t]+)|([;&|\n])|('[^']*'|"[^"\\$`]*"|[^\s'";&|\\$`(){}<>#]+)|([()`])|("(?:[^"\\]|\\[\s\S])*"|\\[\s\S]?|[\s\S])/g;

/** POSIX-quoted words, split into simple commands on ; && || | &, newlines and group edges. */
export function parseShell(command: string): ParsedShell {
  const { lines, odd } = lex(command);
  const commands: SimpleCommand[] = [];
  let unparseable = odd;
  for (const words of lines) {
    const simple = simplify(words);
    unparseable ||= opaque(simple.tokens);
    commands.push(...unwrapKeywords(simple));
  }
  return { commands, unparseable, words: command.match(RAW_WORD) ?? [] };
}

/**
 * Words per simple command, exact for the subset it models: bare words, single quotes
 * (literal), and double quotes holding no escape or expansion. Anything else is read as well as
 * the lexer can and flagged odd.
 */
function lex(command: string): {
  readonly lines: readonly (readonly string[])[];
  readonly odd: boolean;
} {
  const lines: string[][] = [[]];
  let word: string | undefined;
  let odd = false;
  const end = (): void => {
    if (word !== undefined) lines.at(-1)?.push(word);
    word = undefined;
  };
  for (const [, blank, separator, piece, edge, other] of command.matchAll(
    PIECE,
  )) {
    if (piece !== undefined) word = (word ?? "") + unquote(piece);
    else if (blank !== undefined) end();
    else if (separator !== undefined || edge !== undefined) {
      odd ||= edge !== undefined;
      end();
      lines.push([]);
    } else {
      odd = true;
      word = (word ?? "") + readEscaped(other ?? "");
    }
  }
  end();
  return { lines, odd };
}

/** Quotes are removed; the quoted text joins the word. */
function unquote(piece: string): string {
  const q = piece[0];
  return q === "'" || q === '"' ? piece.slice(1, -1) : piece;
}

/** An escape gives its character (a line continuation gives nothing), as the shell reads it. */
function readEscaped(piece: string): string {
  if (piece.startsWith("\\")) return piece.slice(1).replace("\n", "");
  if (piece.length > 1 && piece.startsWith('"'))
    return piece
      .slice(1, -1)
      .replace(/\\([\\$`"\n])/g, (_, c: string) => (c === "\n" ? "" : c));
  return piece;
}

/** Strips leading assignments and wrappers, in any order. */
function simplify(words: readonly string[]): SimpleCommand {
  let rest = words;
  let allowable = true;
  for (;;) {
    const name = ASSIGNMENT.exec(rest[0] ?? "");
    const skip = name === null ? wrapperLength(rest) : 1;
    if (skip === 0) return { tokens: rest, allowable };
    allowable &&= !DANGEROUS_ENV.test(name?.[1] ?? "");
    rest = rest.slice(skip);
  }
}

/**
 * How many leading words are a `timeout [options] N`, `nice`, `nohup` or `time` wrapper,
 * options and a closing `--` included.
 */
function wrapperLength(words: readonly string[]): number {
  const [head] = words;
  if (head === undefined || !WRAPPERS.has(head)) return 0;
  let n = 1;
  for (let w = words[n]; w?.startsWith("-"); w = words[n]) {
    if (w === "--") {
      n += 1;
      break;
    }
    n += OPTIONS_WITH_VALUE.has(w) ? 2 : 1;
  }
  return head === "timeout" ? n + 1 : n;
}

function opaque(tokens: readonly string[]): boolean {
  return (
    OPAQUE.has(tokens[0] ?? "") || tokens.some((w) => w === "{" || w === "}")
  );
}

/**
 * A command that opens with a reserved word, and the command after the word, both as deny and
 * ask rules read them. Neither is allowed.
 */
function unwrapKeywords(simple: SimpleCommand): readonly SimpleCommand[] {
  const [first] = simple.tokens;
  if (first === undefined) return [];
  if (!KEYWORDS.has(first)) return [simple];
  const inner = unwrapKeywords(simplify(simple.tokens.slice(1)));
  return [simple, ...inner].map((c) => ({ ...c, allowable: false }));
}

/** Whitespace-separated words of a rule specifier (quotes removed); none if it isn't modelled. */
export function specTokens(spec: string): readonly string[] {
  const { lines, odd } = lex(spec);
  return odd ? [] : lines.flat();
}
